operators/pkg/custom-metrics-server/metrics_scraper.go

```diff
@@ -0,0 +1,168 @@
+
+		if err := fillCachedMetrics(ctx, log.WithName("fill-cached-metrics")); err != nil {
+			log.Error(err, "Error retrieving containerStats")
+		}
+
+		// Operation running the context completed
+		cancel()
+	}
+}
+
+func fillCachedMetrics(ctx context.Context, log logr.Logger) error {
+	containerIdsMutex.RLock()
+	defer containerIdsMutex.RUnlock()
+
+	cachedMetricsMutex.Lock()
+	defer cachedMetricsMutex.Unlock()
+
+	for i, containerID := range containerIds {
+		log = log.WithValues("containerId", containerID)
+		containerStatsList, err := listContainerStats(ctx, log.WithName("list-container-stats"), &containerIds[i])
+		if err != nil {
+			return err
+		}
+		// If no stats are found for containerID, Pod may no longer be running
+		// and still not being removed from containerIds list.
+		if len(containerStatsList.GetStats()) == 0 {
+			log.V(2).Info("Removing containerId related cached metrics")
+			removeCachedMetric(containerID)
+			continue
+		}
+
+		metrics := &CustomMetrics{}
+		containerStats := containerStatsList.GetStats()[0]
+		metrics.MemBytes = containerStats.GetMemory().GetWorkingSetBytes().GetValue()
+		metrics.DiskBytes = containerStats.GetWritableLayer().GetUsedBytes().GetValue()
+
+		podName := containerStats.Attributes.Labels["io.kubernetes.pod.name"]
+		if oldCPU, ok := oldStats[podName]; ok {
+			duration := containerStats.GetCpu().GetTimestamp() - oldCPU.GetCpu().GetTimestamp()
+			newNs := containerStats.GetCpu().GetUsageCoreNanoSeconds().GetValue()
+			oldNs := oldCPU.GetCpu().GetUsageCoreNanoSeconds().GetValue()
+			cpuPerc := float32(newNs-oldNs) * 100 / float32(duration)
+			metrics.CPUPerc = cpuPerc
+		}
+		oldStats[podName] = containerStats
+		cachedMetrics[podName] = metrics
+	}
+	return nil
+}
+
+func updateApplicationContainerIds(log logr.Logger) {
+	// list of containers watched for stats is updated every updatePeriod.
+	updatePeriod := 5 * time.Second
+
+	ticker := time.NewTicker(updatePeriod)
+	log.V(1).Info("Cycle started", "updatePeriod", updatePeriod)
+	for range ticker.C {
+		if err := findApplicationContainerIds(log.WithName("find-application-container-ids").WithValues("reason", "update cycle"), updatePeriod); err != nil {
+			log.Error(err, "Error retrieving containers")
+		}
+	}
+}
+
+func findApplicationContainerIds(log logr.Logger, responseTimeout time.Duration) error {
+	log = log.WithValues("contextTimeout", responseTimeout)
+	ctx, cancel := context.WithTimeout(context.Background(), responseTimeout)
+	defer cancel()
+
+	wantedState := runtimeapi.ContainerState_CONTAINER_RUNNING
+	containers, err := listContainers(ctx, log.WithName("list-containers"), nil, nil, &wantedState, nil)
+	if err != nil {
+		return err
+	}
+	containerIdsMutex.Lock()
+	defer containerIdsMutex.Unlock()
+	containerIds = []string{}
+	for _, container := range containers {
+		if !utils.Contains(ignoredContainerNames, container.Metadata.Name) &&
+			strings.HasPrefix(container.Labels["io.kubernetes.pod.name"], "instance") {
```

Comment:
There is no guarantee instance names start with instance.

Reply:
The list of watched containerIds is built in 3 steps:
1. List all containerIds running on the Node.
2. Remove from the list non-instance containers.
3. Remove from the remaining Ids sidecars that are not ApplicationContainers (supposing that at most one application container is running in every instance, unique` <podName, containerId>` will remain).

I'll think to a workaround to retrieve the same result, given that not all instance names start with "instance". Any advice?